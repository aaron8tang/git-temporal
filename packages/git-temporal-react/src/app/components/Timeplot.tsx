import React from 'react';
import { style } from 'app/styles';
import * as d3 from 'd3';
require('d3-selection-multi');

import { ICommit } from 'app/interfaces';
import { ZoomContainer } from 'app/components/ZoomContainer';

// https://github.com/wbkd/d3-extended
d3.selection.prototype.moveToFront = function() {
  return this.each(function() {
    // @ts-ignore
    this.parentNode.appendChild(this);
  });
};

export interface TimeplotProps {
  // The children are the menu content
  commits: ICommit[];
  highlightedCommitId?: string;
  style?: string | object;
  onMouseEnter: (evt, date?) => void;
  onMouseLeave: (evt, date?) => void;
  onMouseMove: (evt, date) => void;
  onMouseDown: (evt, date) => void;
  onMouseUp: (evt, date) => void;
}

const timeplotStyle = {
  _extends: 'fill',
};
const blobStyle = {
  fill: '#696969',
  opacity: 0.3,
};

const highlightedBlobStyle = {
  fill: style('selected')['backgroundColor'] || 'blue',
  opacity: 0.8,
};

const LEFT_PADDING = 20;
const PADDING = 20;

export class Timeplot extends React.Component<TimeplotProps> {
  private timeplotRef;

  private svg;
  private xScale;
  private yScale;
  private rScale;

  constructor(props) {
    super(props);
    this.timeplotRef = React.createRef();
    this.getUTCDateOfCommit = this.getUTCDateOfCommit.bind(this);
    this.getHourOfCommit = this.getHourOfCommit.bind(this);
  }

  componentDidMount() {
    this.renderTimeplot();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.commits.length !== this.props.commits.length) {
      this.renderTimeplot();
      return;
    }
    for (let i = 0; i < 20; i++) {
      if (prevProps.commits[i].id !== this.props.commits[i].id) {
        this.renderTimeplot();
        return;
      }
    }
    if (prevProps.highlightedCommitId !== this.props.highlightedCommitId) {
      this.updateHighlightedCommit();
    }
  }

  render() {
    return (
      <ZoomContainer
        style={this.props.style}
        onZoom={() => this.renderTimeplot()}
      >
        <div
          style={style(timeplotStyle)}
          ref={this.timeplotRef}
          onMouseEnter={this.onMouseEnter}
          onMouseLeave={this.onMouseLeave}
          onMouseMove={this.onMouseMove}
          onMouseDown={this.onMouseDown}
          onMouseUp={this.onMouseUp}
        />
      </ZoomContainer>
    );
  }

  private updateHighlightedCommit() {
    this.svg
      .selectAll('circle[data-selected="true"]')
      .styles(style(blobStyle))
      .attr('data-selected', false);
    this.svg
      .selectAll(`circle[data-id="${this.props.highlightedCommitId}"]`)
      .moveToFront()
      .styles(style(highlightedBlobStyle))
      .attr('data-selected', true)
      .transition()
      .duration(500)
      .attr('r', 100)
      .transition()
      .duration(500)
      .attr('r', d => this.rScale(d.linesAdded + d.linesDeleted || 0));
  }

  private clearTimeplot() {
    const element = this.timeplotRef.current;
    while (element.firstChild) {
      element.removeChild(element.firstChild);
    }
  }

  private renderTimeplot() {
    const element = this.timeplotRef.current;
    if (this.props.commits.length <= 0) {
      element.innerHtml =
        "<div class='placeholder'>No commits, nothing to see here.</div>";
      return;
    }
    this.clearTimeplot();

    // this.$element.width(this.zoom * 100 + '%');
    this.svg = d3
      .select(element)
      .append('svg')
      .attr('width', element.clientWidth)
      .attr('height', element.clientHeight);
    this.calculateScales();
    this.renderAxis(this.svg);
    this.renderBlobs(this.svg);
    this.updateHighlightedCommit();
  }

  private getUTCDateOfCommit(commit) {
    const d = new Date(0);
    d.setUTCSeconds(commit.authorDate);
    return d;
  }

  private getHourOfCommit(commit) {
    const d = this.getUTCDateOfCommit(commit);
    return d.getHours();
  }

  private calculateScales() {
    const element = this.timeplotRef.current;
    const { commits } = this.props;
    const w = element.clientWidth;
    const h = element.clientHeight;
    const maxImpact = d3.max(commits.map(d => d.linesAdded + d.linesDeleted));
    const minDate = this.getUTCDateOfCommit(commits[commits.length - 1]);
    const maxDate = this.getUTCDateOfCommit(commits[0]); // Date.now();

    console.log(
      'dates',
      minDate,
      maxDate,
      commits[0].authorDate,
      this.getHourOfCommit(commits[0])
    );

    this.xScale = d3
      .scaleTime()
      .domain([minDate, maxDate])
      .range([LEFT_PADDING, w - PADDING]);
    this.yScale = d3
      .scaleLinear()
      .domain([0, 25])
      .range([10, h - PADDING * 2]);
    this.rScale = d3
      .scalePow(10)
      .domain([1, maxImpact])
      .range([3, 30]);
  }

  private renderAxis(svg) {
    const element = this.timeplotRef.current;
    const h = element.clientHeight;

    const xAxis = d3.axisBottom().scale(this.xScale);
    const yAxis = d3
      .axisLeft()
      .scale(this.yScale)
      .ticks(0);

    svg
      .append('g')
      .attr('class', 'axis')
      .attr('transform', `translate(0, ${h - PADDING})`)
      .call(xAxis);

    svg
      .append('g')
      .attr('class', 'axis')
      .attr('transform', `translate(${LEFT_PADDING - PADDING}, 0)`)
      .call(yAxis);
  }

  private renderBlobs(svg) {
    const { commits } = this.props;
    return svg
      .selectAll('circle')
      .data(commits)
      .enter()
      .append('circle')
      .attr('data-id', d => d.id)
      .attr('cx', d => this.xScale(this.getUTCDateOfCommit(d)))
      .attr('cy', d => this.yScale(this.getHourOfCommit(d)))
      .styles(style(blobStyle))
      .transition()
      .duration(500)
      .attr('r', d => this.rScale(d.linesAdded + d.linesDeleted));
  }

  private callCallbackWithDates(callback, evt) {
    const element = this.timeplotRef.current;
    const rect = element.getBoundingClientRect();
    const relativeLeft = evt.clientX - rect.x + element.scrollLeft;
    const exactDate = this.xScale.invert(relativeLeft);
    const startLeft = relativeLeft - 20 < 0 ? 0 : relativeLeft - 20;
    const startDate = this.xScale.invert(startLeft);
    const endLeft =
      relativeLeft + 20 > element.width ? element.width : relativeLeft + 20;
    const endDate = this.xScale.invert(endLeft);

    callback(evt, { exactDate, startDate, endDate, relativeLeft });
  }

  private onMouseEnter = evt => {
    this.props.onMouseEnter(evt);
  };

  private onMouseLeave = evt => {
    this.props.onMouseLeave(evt);
  };

  private onMouseMove = evt => {
    this.callCallbackWithDates(this.props.onMouseMove, evt);
  };
  private onMouseDown = evt => {
    console.log('onMouseDown', evt);
    this.callCallbackWithDates(this.props.onMouseDown, evt);
  };

  private onMouseUp = evt => {
    this.callCallbackWithDates(this.props.onMouseUp, evt);
  };
}
