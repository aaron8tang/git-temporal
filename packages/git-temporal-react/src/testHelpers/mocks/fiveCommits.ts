// these are the first five commits taken from the node.js test repo used by git-temporal/git-log-scraper
// see, https://raw.githubusercontent.com/git-temporal/git-temporal/master/packages/git-log-scraper/test/snapshots/node.snapshot.json

const fiveCommits = [
  {
    id: 'c698a0f01de1fa682443510920a2dafc2440d6c8',
    authorName: 'Bee Wilkerson',
    relativeDate: '2 days ago',
    authorDate: 1537541919,
    message: 'Explain why this fork exists in README.md',
    body: '',
    hash: 'c698a0f01d',
    linesAdded: 2,
    linesDeleted: 649,
    files: [
      {
        name: 'README.md',
        linesAdded: 2,
        linesDeleted: 649,
      },
    ],
  },
  {
    id: '058c5b81cdbabe8989a194ba5d388f4c230f4af6',
    authorName: 'Tobias Nießen',
    relativeDate: '2 days ago',
    authorDate: 1537272890,
    message: 'crypto: do not allow multiple calls to setAuthTag',
    body:
      'Calling setAuthTag multiple times can result in hard to detect bugssince to the user, it is unclear which invocation actually affectedOpenSSL.PR-URL: https://github.com/nodejs/node/pull/22931Reviewed-By: Anna Henningsen <anna@addaleax.net>Reviewed-By: James M Snell <jasnell@gmail.com>Reviewed-By: Ujjwal Sharma <usharma1998@gmail.com>',
    hash: '058c5b81cd',
    linesAdded: 29,
    linesDeleted: 6,
    files: [
      {
        name: 'doc/api/crypto.md',
        linesAdded: 1,
        linesDeleted: 1,
      },
      {
        name: 'src/node_crypto.cc',
        linesAdded: 2,
        linesDeleted: 5,
      },
      {
        name: 'test/parallel/test-crypto-authenticated.js',
        linesAdded: 26,
        linesDeleted: 0,
      },
    ],
  },
  {
    id: '56493bf1ebfab3ec102fe017f30fa4f81ba6a256',
    authorName: 'Daniel Bevenius',
    relativeDate: '2 days ago',
    authorDate: 1526296167,
    message: 'tools: add bash completion for node',
    body:
      'This commit adds a --completion-bash option to node which can besourced to provide bash code completion for node options.Usage:$ node --completion-bash  > node_bash_completion$ source node_bash_completion$ node --[tab]PR-URL: https://github.com/nodejs/node/pull/20713Reviewed-By: Anna Henningsen <anna@addaleax.net>Reviewed-By: James M Snell <jasnell@gmail.com>Reviewed-By: Denys Otrishko <shishugi@gmail.com>',
    hash: '56493bf1eb',
    linesAdded: 51,
    linesDeleted: 1,
    files: [
      {
        name: 'doc/api/cli.md',
        linesAdded: 11,
        linesDeleted: 0,
      },
      {
        name: 'doc/node.1',
        linesAdded: 3,
        linesDeleted: 0,
      },
      {
        name: 'lib/internal/bash_completion.js',
        linesAdded: 25,
        linesDeleted: 0,
      },
      {
        name: 'lib/internal/bootstrap/node.js',
        linesAdded: 7,
        linesDeleted: 1,
      },
      {
        name: 'node.gyp',
        linesAdded: 1,
        linesDeleted: 0,
      },
      {
        name: 'src/node_options.cc',
        linesAdded: 3,
        linesDeleted: 0,
      },
      {
        name: 'src/node_options.h',
        linesAdded: 1,
        linesDeleted: 0,
      },
    ],
  },
  {
    id: '408e8ce22ce7436e88a9977b4dc99fc3c431fc67',
    authorName: 'Michaël Zasso',
    relativeDate: '3 days ago',
    authorDate: 1537212012,
    message: 'doc: update maintaining V8 guide',
    body:
      'Replace references to the outdated `update-v8` tool to its replacementin `node-core-utils`PR-URL: https://github.com/nodejs/node/pull/22913Reviewed-By: Richard Lau <riclau@uk.ibm.com>Reviewed-By: Colin Ihrig <cjihrig@gmail.com>Reviewed-By: Weijia Wang <starkwang@126.com>Reviewed-By: Sakthipriyan Vairamani <thechargingvolcano@gmail.com>Reviewed-By: Joyee Cheung <joyeec9h3@gmail.com>Reviewed-By: Luigi Pinca <luigipinca@gmail.com>Reviewed-By: James M Snell <jasnell@gmail.com>',
    hash: '408e8ce22c',
    linesAdded: 8,
    linesDeleted: 8,
    files: [
      {
        name: 'doc/guides/maintaining-V8.md',
        linesAdded: 8,
        linesDeleted: 8,
      },
    ],
  },
  {
    id: '54af9738b9fad1c1e46ea9170037146324cd152e',
    authorName: 'Ruben Bridgewater',
    relativeDate: '3 days ago',
    authorDate: 1537269982,
    message: 'doc: specify fast-tracking',
    body:
      'Currently the documentation is not specific how fast-tracking shouldbe applied. This specifies exactly how things should be done to preventconfusion.PR-URL: https://github.com/nodejs/node/pull/22929Reviewed-By: Denys Otrishko <shishugi@gmail.com>Reviewed-By: Anna Henningsen <anna@addaleax.net>Reviewed-By: Tobias Nießen <tniessen@tnie.de>Reviewed-By: Vse Mozhet Byt <vsemozhetbyt@gmail.com>Reviewed-By: John-David Dalton <john.david.dalton@gmail.com>Reviewed-By: Matteo Collina <matteo.collina@gmail.com>Reviewed-By: Luigi Pinca <luigipinca@gmail.com>Reviewed-By: Michaël Zasso <targos@protonmail.com>Reviewed-By: Weijia Wang <starkwang@126.com>Reviewed-By: Richard Lau <riclau@uk.ibm.com>Reviewed-By: Matheus Marchini <mat@mmarchini.me>Reviewed-By: Sakthipriyan Vairamani <thechargingvolcano@gmail.com>Reviewed-By: Yuta Hiroto <hello@hiroppy.me>Reviewed-By: Rich Trott <rtrott@gmail.com>Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>Reviewed-By: George Adams <george.adams@uk.ibm.com>',
    hash: '54af9738b9',
    linesAdded: 6,
    linesDeleted: 3,
    files: [
      {
        name: 'COLLABORATOR_GUIDE.md',
        linesAdded: 6,
        linesDeleted: 3,
      },
    ],
  },
];

export default fiveCommits;
