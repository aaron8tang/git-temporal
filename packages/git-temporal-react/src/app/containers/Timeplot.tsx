import React from 'react';
import { connect } from 'react-redux';
import { delay } from 'lodash';
import { debug } from '@git-temporal/logger';

import { style } from 'app/styles';
import { ITimeplotState, ICommit, DispatchProps } from 'app/interfaces';

import { getTimeplotContainerState } from 'app/selectors';
import { setDates } from 'app/actions/setDates';

import { debounce } from 'app/utilities/debounce';
import { throttle } from 'app/utilities/throttle';
import { ZoomContainer, ICustomZoom } from 'app/components/ZoomContainer';
import { TimeplotGraph } from 'app/components/TimeplotGraph';
import { EpochSpan } from 'app/components/EpochSpan';
import { CommaNumber } from 'app/components/CommaNumber';

import {
  TimeplotPopup,
  TIMEPLOT_POPUP_WIDTH,
} from 'app/components/TimeplotPopup';

import { filterCommitsForSpan } from 'app/utilities/commits';

interface TimeplotLocalState {
  hoverMarkerLeft: number;
  scrollLeft: number;
  popupOpen: boolean;
  timeplotRenders: number;
  popupCommits: ICommit[];
  popupStartDate?: Date;
  popupEndDate?: Date;
  customZooms: ICustomZoom[];
  mouseInPopover: boolean;
}

const initialState = {
  // TODO: this should be a constant. I think it's used elsewhere?
  hoverMarkerLeft: -40, // -40 = out of sight
  scrollLeft: 0,
  popupOpen: false,
  timeplotRenders: 0,
  popupCommits: [],
  popupStartDate: null,
  popupEndDate: null,
  customZooms: [],
  mouseInPopover: false,
};

const outerStyle = {
  overflow: 'visible', // needed for popup
  flex: '0 0 200px',
};

const graphContainerStyle = {
  _extends: ['borderedPanel', 'flexColumn'],
  flexGrow: 0,
  flexShrink: 0,
  position: 'relative',
  marginTop: 10,
  minHeight: 100,
};

const statsStyle = {
  _extends: 'normalText',
  width: '100%',
  textAlign: 'center',
};

const timeplotStyle = {
  _extends: 'fill',
};

const markerStyle = {
  position: 'absolute',
  height: 130,
  width: 10,
  opacity: 0.5,
  zIndex: 1,
  top: 0,
};

const hoverMarkerStyle = {
  _extends: markerStyle,
  backgroundColor: '@colors.selectable',
};

export class Timeplot extends React.Component<
  ITimeplotState & DispatchProps,
  TimeplotLocalState
> {
  readonly state: TimeplotLocalState = initialState;
  private timeplotRef: any;
  private outerRef: any;
  private debouncedOnMouseLeave: (args: any) => void;
  private debouncedOnMouseMove: (args: any) => void;

  // we don't want to necessarily rerender on changes to these
  private lastMouseMoveCoords: { pageX: number; pageY: number };
  private lastMouseDownDate: Date;
  private lastRerenderRequestedAt: Date;

  constructor(props) {
    super(props);
    this.lastRerenderRequestedAt = Date.now() as any;
    this.timeplotRef = React.createRef();
    this.outerRef = React.createRef();
    this.debouncedOnMouseLeave = debounce(this.onMouseLeave, 100);
    this.debouncedOnMouseMove = throttle(this.onMouseMove, 100);
  }

  componentDidUpdate(prevProps, prevState) {
    if (!this.timeplotRef.current) {
      return;
    }
    this.timeplotRef.current.focus();
    if (
      prevProps.startDate !== this.props.startDate ||
      prevProps.endDate !== this.props.endDate
    ) {
      this.addCustomZooms();
    }
    if (this.lastRerenderRequestedAt < this.props.rerenderRequestedAt) {
      debug('forcing timeplot rerender');
      this.lastRerenderRequestedAt = this.props.rerenderRequestedAt;
      this.setState({ timeplotRenders: this.state.timeplotRenders + 1 });
    }
  }

  render() {
    const { commits = [], startDate, endDate } = this.props;
    if (!commits || commits.length === 0) {
      return null;
    }
    const outerLeft =
      (this.outerRef.current &&
        this.outerRef.current.getBoundingClientRect().x) ||
      0;
    const popupLeft =
      this.state.hoverMarkerLeft - this.state.scrollLeft <
      TIMEPLOT_POPUP_WIDTH + 20
        ? this.state.hoverMarkerLeft - this.state.scrollLeft + outerLeft - 10
        : this.state.hoverMarkerLeft -
          this.state.scrollLeft -
          TIMEPLOT_POPUP_WIDTH +
          outerLeft;

    const firstCommitTime = commits[commits.length - 1].authorDate;
    const lastCommitTime = commits[0].authorDate;

    return (
      <div style={style(outerStyle)} ref={this.outerRef}>
        <div style={style(graphContainerStyle)}>
          <ZoomContainer
            onZoom={this.onZoom}
            onMouseLeave={this.debouncedOnMouseLeave}
            onScroll={this.onScroll}
            customZooms={this.state.customZooms}
            scrollLeft={this.state.scrollLeft}
          >
            <TimeplotGraph
              forceRender={this.state.timeplotRenders}
              commits={this.props.commits}
              style={style(timeplotStyle)}
              ref={this.timeplotRef}
              highlightedCommitIds={this.props.highlightedCommitIds}
              startDate={startDate}
              endDate={endDate}
              onMouseMove={this.debouncedOnMouseMove}
              onMouseDown={this.onMouseDown}
              onMouseUp={this.onMouseUp}
            />
            <div
              style={style(hoverMarkerStyle, {
                left: this.state.hoverMarkerLeft,
              })}
              onMouseMove={this.onMouseHoverMarker}
              onMouseEnter={this.onMouseHoverMarker}
            />
          </ZoomContainer>

          <div style={style(statsStyle)}>
            <CommaNumber value={this.props.commits.length} /> commits spanning{' '}
            <EpochSpan
              firstEpochTime={firstCommitTime}
              secondEpochTime={lastCommitTime}
            />
          </div>
        </div>
        <TimeplotPopup
          commits={this.state.popupCommits}
          isOpen={this.state.popupOpen}
          startDate={this.state.popupStartDate}
          endDate={this.state.popupEndDate}
          left={popupLeft}
          onClose={this.onPopupClose}
          onCommitSelected={this.onCommitSelected}
          onMouseEnter={this.onMouseEnterPopup}
          onMouseLeave={this.onMouseLeavePopup}
        />
      </div>
    );
  }

  private onPopupClose = () => {
    this.setState({ popupOpen: false });
  };

  private onScroll = scrollLeft => {
    this.setState({ scrollLeft });
  };

  private onZoom = () => {
    this.setState({
      timeplotRenders: this.state.timeplotRenders + 1,
    });
  };

  private onMouseEnterPopup = evt => {
    this.setState({ mouseInPopover: true });
  };

  private onMouseLeavePopup = evt => {
    this.setState({ mouseInPopover: false, popupOpen: false });
  };

  private onMouseLeave = _evt => {
    delay(() => {
      if (!this.state.mouseInPopover) {
        this.setState({ popupOpen: false });
      }
    }, 250);
  };

  private onMouseMove = (evt, { startDate, endDate, relativeLeft }) => {
    const popupCommits = filterCommitsForSpan(
      this.props.commits,
      startDate,
      endDate
    );
    const { pageX, pageY } = evt;
    // this enables greater agility, if the users vertical movement is
    // greater than their horizontal, they might be headed to the popup
    if (!this.lastMouseMoveCoords || pageY >= this.lastMouseMoveCoords.pageY) {
      this.setState({
        popupCommits,
        popupOpen: true,
        popupStartDate: startDate,
        popupEndDate: endDate,
        // +3 : the marker needs to not get in the way of clicking the graph underneath
        //    otherwise you end up clicking on the marker itself
        hoverMarkerLeft: relativeLeft + 3,
      });
    }
    this.lastMouseMoveCoords = { pageX, pageY };
  };

  private onMouseDown = (evt, { startDate }) => {
    debug('Timeplot: onMouseDown', evt.shiftKey, startDate);
    evt.preventDefault();
    this.lastMouseDownDate = startDate;
    this.setDates(startDate, null);
  };

  private onMouseUp = (evt, { startDate }) => {
    debug(
      'Timeplot: onMouseUp',
      evt.shiftKey,
      startDate,
      this.lastMouseDownDate
    );
    evt.preventDefault();
    if (!this.lastMouseDownDate) {
      return;
    }
    if (startDate !== this.lastMouseDownDate) {
      this.setDates(startDate, this.lastMouseDownDate);
    } else {
      this.setDates(startDate, null);
    }
    this.lastMouseDownDate = null;
  };

  private onMouseHoverMarker = evt => {
    this.setState({
      hoverMarkerLeft:
        this.state.hoverMarkerLeft + (evt.pageX - evt.clientX) + 3,
    });
  };

  private onCommitSelected = (evt, commit) => {
    evt.stopPropagation();
    const epochAuthorDate = commit.authorDate * 1000;
    // console.log('onCommitSelected', evt, commit, this.props.startDate);
    // TODO: test: you should be able to isolate a single commit (`+ 1` below)
    // if the user clicks the same commit twice we select just that commit
    if (epochAuthorDate === this.props.startDate) {
      this.setDates(epochAuthorDate, epochAuthorDate + 1);
    } else {
      this.setDates(epochAuthorDate, null);
    }
  };

  private setDates(startDate, endDate) {
    const { dispatch } = this.props;
    dispatch(setDates(startDate, endDate));
  }

  private addCustomZooms() {
    const { startDate, endDate } = this.props;
    if (!startDate || !this.timeplotRef.current) {
      this.setState({ customZooms: [] });
      return;
    }
    const { xScale, getScrollWidth } = this.timeplotRef.current;
    const defaultedEndDate = endDate ? endDate * 1000 : new Date();
    const spanLeft = xScale(startDate * 1000);
    const spanRight = xScale(defaultedEndDate);
    const span = spanRight - spanLeft + 6;
    if (span <= 0) {
      this.setState({ customZooms: [] });
      return;
    }
    const scrollWidth = getScrollWidth();
    this.setState({
      customZooms: [
        {
          value: Math.floor((scrollWidth / span) * 100),
          label: 'Zoom to selected time span',
        },
      ],
    });
  }
}

export default connect(getTimeplotContainerState)(Timeplot);
