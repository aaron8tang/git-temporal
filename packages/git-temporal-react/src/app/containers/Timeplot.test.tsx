import React from 'react';
import { mount } from 'enzyme';

import fiveCommits from 'testHelpers/mocks/fiveCommits';
import { Timeplot } from './Timeplot';

describe('components/Timeplot', () => {
  describe('when rendered', () => {
    let wrapper;
    beforeAll(() => {
      wrapper = mount(
        <Timeplot
          selectedPath="some/path/to/nowhere"
          highlightedCommitIds={[fiveCommits[2].id]}
          commits={fiveCommits}
          totalCommits={5}
          authors={6}
          isFetching={false}
          startDate={fiveCommits[3].authorDate}
          endDate={fiveCommits[2].authorDate}
          earliestCommitDate={fiveCommits[0].authorDate}
          latestCommitDate={fiveCommits.slice(-1)[0].authorDate}
          rerenderRequestedAt={null}
          dispatch={jest.fn()}
        />
      );
    });

    test('it should match snapshot (it should have rendered a block element for each line with <br><br>")', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('and then changing the startDate', () => {
      wrapper.setProps({ startDate: fiveCommits[4].authorDate + 1 });
    });

    test('onMouseDown with shiftKey should set dates', () => {
      wrapper
        .instance()
        .onMouseDown(
          { shiftKey: true, preventDefault: jest.fn() },
          { startDate: new Date(fiveCommits[3].authorDate * 1000) }
        );
    });
    test('onMouseUp with shiftKey should set dates', () => {
      wrapper
        .instance()
        .onMouseDown(
          { shiftKey: false, preventDefault: jest.fn() },
          { startDate: new Date(fiveCommits[4].authorDate * 1000) }
        );
    });

    test('onMouseUp without shiftKey should set dates', () => {
      wrapper
        .instance()
        .onMouseDown(
          { shiftKey: false, preventDefault: jest.fn() },
          { startDate: new Date(fiveCommits[3].authorDate * 1000) }
        );
    });
  });
});
