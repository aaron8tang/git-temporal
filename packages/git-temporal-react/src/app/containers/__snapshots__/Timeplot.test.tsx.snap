// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Timeplot when rendered with mock redux store should match snapshot 1`] = `
<body>
  <div>
    <div
      style="overflow: visible; flex: 0 0 120px;"
    >
      <div
        style="background: whitesmoke; border-radius: 3px; color: rgb(51, 51, 51); margin-right: 7px; padding: 7px; display: flex; flex-direction: column; flex: 0 0 120px; min-height: 120px; position: relative; margin-top: 5px; overflow: hidden;"
      >
        <div
          style="display: block; height: 100%; width: 100%; position: relative;"
        >
          <div
            style="position: relative; height: 100%; width: 100%;"
          >
            <div
              style="display: flex; width: 100%; position: relative;"
            >
              <div
                data-testid="outerContainer"
                style="position: relative; height: 100%; width: 100%; overflow-x: scroll;"
              >
                <div
                  data-testid="innerContainer"
                  style="position: relative; display: block;"
                >
                  <div
                    style="height: 100%; width: 100%;"
                  >
                    <div
                      data-testid="timeplotGraph"
                      style="position: relative; height: 100%; width: 100%; overflow: hidden; background: rgb(254, 254, 254); color: rgb(51, 51, 51);"
                    >
                      <div
                        style="position: relative; height: 100%; width: 100%;"
                      >
                        <svg
                          height="120"
                          width="0"
                        >
                          <g
                            class="axis"
                            fill="none"
                            font-family="sans-serif"
                            font-size="10"
                            text-anchor="middle"
                            transform="translate(0, 100)"
                          >
                            <path
                              class="domain"
                              d="M20.5,6V0.5H-19.5V6"
                              stroke="currentColor"
                              style="stroke: #333333;"
                            />
                          </g>
                          <circle
                            cx="-20"
                            cy="38"
                            data-id="94932bedc09976aeac43615d95906aa2dcfefa1f"
                            style="fill: #696969; opacity: 0.2;"
                          />
                          <circle
                            cx="-13.627132231902591"
                            cy="52"
                            data-id="c03bb366e0e901b33acd9783db3003cb49eddf80"
                            style="fill: #696969; opacity: 0.2;"
                          />
                          <circle
                            cx="4.386196753652295"
                            cy="60"
                            data-id="1a66971f9e2f255d6278ade61e70517faf5839db"
                            style="fill: #696969; opacity: 0.2;"
                          />
                          <circle
                            cx="4.635503644658122"
                            cy="36"
                            data-id="5d24a5e88f67ad303e2169795f080009241f9f86"
                            style="fill: #696969; opacity: 0.2;"
                          />
                          <circle
                            cx="12.363685655618129"
                            cy="66"
                            data-id="211694a8a22a4c4b954e729e0d7eb946cf0f8d60"
                            style="fill: #696969; opacity: 0.2;"
                          />
                          <circle
                            cx="13.419230076697955"
                            cy="62.00000000000001"
                            data-id="8a1ee2f7a531d83cb76cc882d1e346c5cdd55f56"
                            style="fill: #696969; opacity: 0.2;"
                          />
                          <circle
                            cx="17.500973741415585"
                            cy="57.99999999999999"
                            data-id="2de57fe15a69a84fadb08ce549599c7de5a9d3b2"
                            style="fill: #696969; opacity: 0.2;"
                          />
                          <circle
                            cx="20"
                            cy="60"
                            data-id="4c0c81953d090dbf122e126f2b6b8c803d4a0569"
                            style="fill: #696969; opacity: 0.2;"
                          />
                        </svg>
                      </div>
                    </div>
                    <div
                      style="position: absolute; height: 120px; width: 10px; opacity: 0.5; z-index: 1; top: 0px; background-color: lightskyblue; left: -40px;"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              style="position: absolute; right: 18px; background: transparent; top: 3px; z-index: 1;"
            >
              <label
                style="font-size: 12px; line-height: 1.35; font-weight: normal;"
              >
                <select>
                  <option
                    value="100"
                  >
                    100% zoom
                  </option>
                  <option
                    value="200"
                  >
                    200% zoom
                  </option>
                  <option
                    value="300"
                  >
                    300% zoom
                  </option>
                  <option
                    value="500"
                  >
                    500% zoom
                  </option>
                  <option
                    value="800"
                  >
                    800% zoom
                  </option>
                  <option
                    value="1300"
                  >
                    1300% zoom
                  </option>
                </select>
              </label>
            </div>
          </div>
          <div
            style="font-size: 12px; line-height: 1.35; font-weight: normal; width: 100%; text-align: center; padding-top: 3px; padding-bottom: 3px; position: absolute; top: 7px; text-shadow: 2px 2px #FEFEFE;"
          >
            <span
              style="opacity: 0.8; background-color: rgb(254, 254, 254);"
            >
              Total of 
              <span>
                8
              </span>
              <span>
                 commits by 
              </span>
              <span>
                7
              </span>
              <span>
                 authors spanning 
              </span>
              <span>
                3 months
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
